### PR TITLE
Notification feature : Feedback changes

### PR DESCRIPTION
**Pull request by alkeshrails**
https://aicrowd.monday.com/boards/363093650/pulses/610004882

### CODE REVIEW

File: app/services/notification_service.rb [hunk 1]
```diff
@@ -44,19 +45,21 @@ def failed
         message:           message,
         thumbnail_url:     thumb,
         notification_url:  link,
+        challenge_id:      @notifiable.challenge.id,
         is_new:            true)
   end
 
   def leaderboard
     return if @participant.nil?
```
Comment (skbly7) [inline app/services/notification_service.rb:53]: return if @notifiable.previous_row_num == @notifiable.row_num

File: app/controllers/participants_controller.rb [hunk 2]
```diff
@@ -102,8 +102,15 @@ def notifications_message
   end
 
   def read_notification
-    @notification = current_user.notifications.find(params[:id])
-    @notification.update(is_new: false)
+    if params[:ids].present?
+      @notifications = current_user.notifications.find(JSON.parse(params[:ids]))
+      @notifications.each do |notification|
```
Comment (piotrekpasciak) [inline app/controllers/participants_controller.rb:107]: Use either `update!` to make sur everything works or `update_all` to make action faster. It doesn't make sense to validate objects and do nothing about validations.

File: app/controllers/participants_controller.rb [hunk 3]
```diff
@@ -102,8 +102,15 @@ def notifications_message
...
+        notification.update(is_new: false)
+      end
+    else
+      @notification = current_user.notifications.find(params[:id])
+      @notification.update(is_new: false)
```
Comment (piotrekpasciak) [inline app/controllers/participants_controller.rb:112]: `update!`

File: app/services/notification_service.rb [hunk 4]
```diff
@@ -28,13 +32,18 @@ def graded
...
         is_new:            true)
   end
 
   def failed
     message = "Your #{@notifiable.challenge.challenge} Challenge submission ##{@notifiable.id} failed to evaluate."
     thumb   = @notifiable.challenge.image_file.url
-    link    = challenge_submissions_url(@notifiable.challenge, my_submissions: true)
+    link    = challenge_submission_url(@notifiable.challenge, @notifiable.id)
+
+    existing_notification = @participant.notifications.where(notification_type: "failed", notifiable_type: "Submission", message: message, is_new: true).first
```
Comment (skbly7) [inline app/services/notification_service.rb:44]: The message and type doesn't really need to be same for returning.

Can we check if *any* `notifiable_type: "Submission"` is present for this submission ID and return if true?
(same for graded function above)

**Pull request merged**